### PR TITLE
[sipify] implement template inheritance

### PR DESCRIPTION
**Pull request by 3nids**
no need to copy template methods to sip implementations anymore!

### CODE REVIEW

**Comment by 3nids** [comment]
@nyalldawson I don't understand why the sip coverage test fail. QgsOptional is now added to the bindings and documented. Any idea?

**Comment by nyalldawson** [comment]
But how can QgsOptional be used from python? It's a c++ template class and has no meaning there. I think you'll need to bring back the note about using QgsOptionalExpression in python

**Comment by 3nids** [comment]
yep, you cannot use it from Python, will put back the comment.
Although I have to keep the file in SIP otherwise I get `Super-class list contains an invalid type`.
Still any idea why the sip doc coverage test failed despite the presence of qgsoptional.sip?

**Comment by nyalldawson** [comment]
The sip coverage test checks for the actual availability of each method/class in the build. So it doesn't consider the sip files at all, just the end result...

**Comment by 3nids** [comment]
so the issue is that you can't call QgsOptional from python?
then I should add it to the acceptable missing doc file?

**Comment by nyalldawson** [comment]
It just needs the "not available in python bindings" note readded. And that's correct - it's not available in the bindings. We still need it to be checked for documentation coverage though.

File: src/core/qgsoptional.h [hunk 1]
```diff
@@ -28,9 +28,8 @@
  * more internal configuration information that should not be lost when disabling and re-enabling.
  *
  * \since QGIS 3.0
- * \note For Python you need to use implementations for specific template classes
```
Comment (nyalldawson) [inline src/core/qgsoptional.h]: You just need to revert the removal of this line

**Pull request merged**